Support the withdraw zero trick

ch1bo opened the pull request:
Closes #1795

We can mock the ledger's certificate state with reward accounts for all withdrawals with 0 value in a transaction. This will allow included scripts with Rewarding script purpose to run once for the transaction.

The idea of mocking the certificate state emerged in a discussion with @colll78.

We can remove this once CIP-112 is implemented in the cardano-ledger.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated
* [x] Documentation updated -> new How-to
* [x] Haddocks updated
* [x] No new TODOs introduced

ch1bo committed:
Add an e2e scenario to test withdraw 0 trick

ch1bo committed:
Support withdraw-zero trick transactions

We can mock the ledger's certificate state with reward accounts for all
withdrawals with 0 value in a transaction. This will allow included
scripts with Rewarding script purpose to run once for the transaction.

We can remove this once CIP-112 is implemented in the cardano-ledger.

ch1bo committed:
Add a rewarding dummy validator

This proves that the withdraw zero implementation works, but a dedicated
script and test suite is needed before merging this.

noonio commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -35,58 +38,14 @@ import Data.ByteString (isInfixOf)
-  pattern TxOut,
-  pattern TxOutDatumNone,
- )
+import Hydra.Cardano.Api (Coin (..), Era, File (File), Key (SigningKey), PaymentKey, Tx, TxId, UTxO, addTxIns, defaultTxBodyContent, fromLedgerTx, getTxBody, getTxId, getVerificationKey, lovelaceToValue, makeSignedTransaction, mkScriptAddress, mkScriptDatum, mkScriptWitness, mkTxOutDatumHash, mkVkAddress, scriptWitnessInCtx, selectLovelace, signTx, toLedgerData, toLedgerExUnits, toLedgerScript, t ...
```
what on earth happened here? it seems like this is even well-formatted; ( i.e. treefmt doesn't change it ) but why/how?!

ch1bo replied:
HLS sometimes inlines and fourmolu does honor single line inputs. It's not in the diff anymore.

ch1bo committed:
Render errors in evaluateTx more pretty

ch1bo committed:
Sketch how-to use the withdraw zero trick

ch1bo committed:
Update protocol-parameters used by demo and tests

The one we had was severely outdated. This is the one from preview
testnet with fees and prices set to 0

ch1bo committed:
Update withdraw zero tutorial

The dummy validator we use must also be able to certify if we want to
demonstrate the (not needed!) registration step.

ch1bo committed:
Update changelog

noonio commented on docs/docs/how-to/withdraw-zero.md:
```diff
@@ -0,0 +1,82 @@
+  --signing-key-file hydra-cluster/config/credentials/alice-funds.sk \
+  --out-file tx-signed.json
+
+cat tx-signed.json | jq -c '{tag: "NewTx", transaction: .}' | websocat ws://localhost:4001
```
Note, one can _almost_ copy and paste these commands, but you need to have a note in here to run an offline head with the utxo.json you created also in this PR; i.e.

```json utxo.json
{
  "0000000000000000000000000000000000000000000000000000000000000000#0": {
    "address": "addr_test1vqg9ywrpx6e50uam03nlu0ewunh3yrscxmjayurmkp52lfskgkq5k",
    "address": "addr_test1vp5cxztpc6hep9ds7fjgmle3l225tk8ske3rmwr9adu0m6qchmx5z",
    "value": {
      "lovelace": 100000000
    }
  }
}
```

and then

```
cabal exec hydra-node -- \
  --offline-head-seed 0001 \
  --initial-utxo utxo.json \
  --hydra-signing-key demo/alice.sk \
  --ledger-protocol-parameters hydra-cluster/config/protocol-parameters.json
```

ch1bo replied:
Okay I'll add it to a collapsible section

ch1bo replied:
https://github.com/cardano-scaling/hydra/commit/68d477d5e8da4000a9553a1f3f47f550ac9963ff

noonio commented on docs/docs/how-to/withdraw-zero.md:
```diff
@@ -0,0 +1,82 @@
+
+The Hydra L2 ledger is mostly _isomorphic_ to the L1 Cardano ledger. This means that it honors all the same ledger rules and transactions use the exact same format. However, some things related to the Cardano Proof-of-Stake consensus are represented different because they would not make much sense on an L2. For exapmle, transactions that [register stake certificates](https://docs.cardano.org/deve ...
+
+## Rewarding script
```
It might be worth saying something about _why_ you'd want this; i.e. the custom ledger + fanout or not use-case.

ffakenz replied:
Agree.

This trick looks great, but I still don’t get why/when/where I would want to use it.

ch1bo replied:
Isn't "run a script once per transaction" enough of a purpose? There are as many reasons for doing that as for submitting normal transactions or spending scripts in a Hydra head. Our users also don't give more rationale than "to run stuff only once" as seen in https://github.com/cardano-scaling/hydra/issues/1757#issuecomment-2603432940

noonio replied:
@ch1bo Sure, it's fine enough I suppose, but more what I meant is me myself I wouldn't _immediately_ know this is a solution to the "how to make L2 states finalisable when using a custom ledger".

There could be a section in our docs like "cookbook" or "common patterns" where we document these tips and tricks.

noonio reviewed (changes requested):
Looks great; happy to merge with some minor changes to the docs.

Thanks!

ffakenz commented on docs/docs/how-to/withdraw-zero.md:
```diff
@@ -0,0 +1,82 @@
+---
+sidebar_position: 2
```
minor: i think this will conflict with `how-to/submit-transaction`

ch1bo replied:
Good point, I'll update it.

ch1bo replied:
https://github.com/cardano-scaling/hydra/commit/a59270c417c6255e98b825245427b3e4807aed36

ffakenz commented on docs/docs/how-to/withdraw-zero.md:
```diff
@@ -0,0 +1,82 @@
+  --tx-out addr_test1vp5cxztpc6hep9ds7fjgmle3l225tk8ske3rmwr9adu0m6qchmx5z+100000000 \
+  --withdrawal stake_test17zadf9dcekqn9cxkg3q68y56f4d6ujxsv8l6kq4kz2cuduc3gm35e+0 \
+  --withdrawal-script-file rewarding.plutus \
+  --withdrawal-redeemer-value "{}" \
```
question: does this represents the zero value?

ch1bo replied:
I think this translates to a Constr 0 [] in plutus core. It's irrelevant in our example as we do not access the redeemer in the script code (see dummyRewardingScript)

ch1bo committed:
Update order in how-to section

Also remove the not-maintained "etcd" how-to.

ch1bo committed:
Add offline head instructions to withdraw zero how-to
